File: Java/D11/d11p1.java
/*
There is a set of 'N' bags contains apples.
Each bag carries few number of apples.
You need to pick 'K' bags everytime from left to right.
and find the bag which contains max number of apples 'M'.
You need to return the value fo M each time.

Bags array indicates the apples count in each bag.

NOTE:
-----
Can you solve it in linear time?

Input Format:
-------------
Line-1: Two space separated integers, N and K
Line-2: N space separated integers, 

Output Format:
--------------
Print the (N-K+1) integer output.


Sample Input-1:
---------------
8 3
1 2 3 5 4 6 8 7

Sample Output-1:
----------------
3 5 5 6 8 8

Explanation:
------------

K Bags		Max
-----------    -----
1 2 3			3
2 3 5			5
3 5 4			5
5 4 6			6
4 6 8			8
6 8 7			8

*/

/**
 * d11p1
 */

import java.util.*;

// Time Complexity: O(n)
/**
 * d11p1
 */
public class d11p1 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int k = sc.nextInt();
        int[] bags = new int[n];
        for (int i = 0; i < n; i++) {
            bags[i] = sc.nextInt();
        }
        sc.close();
        List<Integer> window = new ArrayList<>();
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < n; i++) {
            if (i < k) {
                max = Math.max(max, bags[i]);
                window.add(bags[i]);
            } else {
                System.out.print(max + " ");
                window.remove(0);
                window.add(bags[i]);
                max = Math.max(max, bags[i]);
                if (window.contains(max) == false) {
                    max = Collections.max(window);
                }
            }
        }
        System.out.println(max);
    }
}

File: CPP/D11/d11p1.cpp
// Sam and Lam want to distribute milk to n houses in their area. The houses are 
// arranged in a row with label from 0 to n - 1 from left to righ,
// where the i-th house is located at x = i.

// Each house needs a specific amount of milk. Sam and Lam have a their own milk 
// and initially full. They distribute the milk in the following way:
//     - Sam gives milk to the houses in order from left to right, starting 
//     from the 0th house. Lam gives milk in order from right to left, starting 
//     from (n-1)th house. 
    
//     - They begin distributing the milk simultaneously. It takes the same amount
//     of time to give milk to each house regardless of how much milk it needs.

//     - Sam/Lam must distribute milk, if they have enough in their can to fully 
//     give it. Otherwise, they first refill their can (instantaneously), then 
//     give the milk to the house. In case both Sam and Lam reach the same house, 
//     the one with more milk currently in their milk can should give milk to the 
//     house. If they have the same amount of milk, then Sam should give the milk.

// Given array of house needs starting from 0 to N-1 integers, where house[i] 
// is the amount of milk the i-th house needs, and two integers capA and capB 
// representing the capacities of Sam's and Lam's milk cans respectively. Return
// the number of times they have to refill to distribute milk to all the houses.


// Input Format:
// -------------
// Line-1: 3 space separarted integers, indiactes N, T and V.
// Line-2: N space separated integers, screen-plays[].

// Output Format:
// --------------
// Print an integer result.


// Sample Input:
// -------------
// 4 5 5
// 5 3 3 4

// Sample Output: 
// --------------
// 2

// Explanation:
// ------------
// - Initially, Sam and Lam have 5 units of milk each in their cans.
// - Sam gives milk to house 0, Lam gives to house 3.
// - Sam and Lam now have 0 units and 1 unit of milk respectively.
// - Sam has to refil can for house 1, and Lam also does not have enough milk 
// for house 3, so he refills his can. So, the total number of times they have
// to refill  0 + 0 + 1 + 1 = 2.


// Sample Input:
// -------------
// 3 9 7
// 4 5 2

// Sample Output: 
// --------------
// 0



#include<bits/stdc++.h>
using namespace std;

int main(){
    int n,a,b;
    cin>>n>>a>>b;
    int a1=a,b1=b;
    vector<int> nums(n,0);
    for(int i=0;i<n;++i){
        cin>>nums[i];
    }
    int i=0,j=n-1;
    int ans = 0;
    while(i<=j){
        if(i==j){
            if(nums[i] > a1 && nums[i]>b1){
                ++ans;
                break;
            }
            break;
        }
        if(nums[i] <= a1){
            a1 -= nums[i];
            i++;
        }
        if(nums[j] <= b1){
            b1 -= nums[j];
            --j;
        }
        if(i==j)continue;
        if(i>j)break;
        if(nums[i] > a1 && nums[i]<=a){
            a1 = a;
            ++ans;
        }
        if(nums[j] > b1 && nums[j]<=b){
            b1 = b;
            ++ans;
        }
    }
    cout<<ans<<endl;
}